// hnpwa-api/api/user.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const moment = require("moment");
/**
 * Get a user by their id and transform into a UI friendly JSON object.
 * @param id - The user's id
 */
function getUser(id, firebaseApp) {
    return __awaiter(this, void 0, void 0, function* () {
        const userRef = firebaseApp.database().ref('v0').child('user').child(id.toString());
        const snapshot = yield userRef.once('value');
        const user = snapshot.val();
        if (user && user.id) {
            return {
                about: user.about,
                created_time: user.created,
                created: moment(user.created * 1000).fromNow(),
                id: user.id,
                karma: user.karma
            };
        }
        return null;
    });
}
exports.getUser = getUser;
